cabal-install/Distribution/Client/SetupWrapper.hs

```diff
@@ -389,26 +389,33 @@ externalSetupMethod verbosity options pkg bt mkargs = do
       Nothing    -> getInstalledPackages verbosity
                     comp (usePackageDB options') conf
 
-  cabalLibVersionToUse :: IO (Version, (Maybe ComponentId)
+  cabalLibVersionToUse :: IO (Version, Maybe ComponentId
                              ,SetupScriptOptions)
   cabalLibVersionToUse =
-    case useCabalSpecVersion options of
-      Just version -> do
+    case find (hasCabal . snd) (useDependencies options) of
```

Comment:
@23Skidoo I agree that the logic is becoming complicated.  We could separate out the old and new code paths, as suggested here https://github.com/haskell/cabal/pull/3094#issuecomment-177631386.  Then it would be easier to see which `SetupScriptOptions` apply in each case.